src/vm/functions/database.rs

```diff
@@ -8,34 +8,97 @@ use vm::representations::{SymbolicExpression, SymbolicExpressionType};
     check_arguments_at_least(2, args)?;
 
-    let contract_identifier = match args[0].expr {
-        SymbolicExpressionType::LiteralValue(Value::Principal(PrincipalData::Contract(ref contract_identifier))) => contract_identifier,
-        _ => return Err(CheckErrors::ContractCallExpectName.into())
-    };
-
     let function_name = args[1].match_atom()
         .ok_or(CheckErrors::ExpectedName)?;
+    let rest_args: Vec<_> = {
+        let rest_args = &args[2..];
+        let rest_args: Result<Vec<_>> = rest_args.iter().map(|x| { eval(x, env, context) }).collect();
+        let mut rest_args = rest_args?;
+        rest_args.drain(..).map(|x| { SymbolicExpression::atom_value(x) }).collect()
+    };
 
-    let rest_args = &args[2..];
-
-    let rest_args: Result<Vec<_>> = rest_args.iter().map(|x| { eval(x, env, context) }).collect();
-    let mut rest_args = rest_args?;
-    let rest_args: Vec<_> = rest_args.drain(..).map(|x| { SymbolicExpression::atom_value(x) }).collect();
+    let (contract_identifier, type_returns_constraint) = match &args[0].expr {
+        SymbolicExpressionType::LiteralValue(Value::Principal(PrincipalData::Contract(ref contract_identifier))) => {
+            // Static dispatch
+            (contract_identifier, None)
+        },
+        SymbolicExpressionType::Atom(contract_ref) => {
+            // Dynamic dispatch
+            match context.callable_contracts.get(contract_ref) {
+                Some((ref contract_identifier, trait_identifier)) => {
+                    let contract_to_check = env.global_context.database.get_contract(contract_identifier)
+                        .map_err(|_e| CheckErrors::NoSuchContract(contract_identifier.to_string()))?;
+                    let contract_context_to_check = contract_to_check.contract_context;
+
+                    // Attempt to short circuit the dynamic dispatch checks:
+                    // If the contract is explicitely implementing the trait with `impl-trait`,
+                    // then we can simply rely on the analysis performed at publish time.
+                    if contract_context_to_check.is_explicitely_implementing_trait(&trait_identifier) {
+                        (contract_identifier, None)
+                    } else {
+                        let trait_name = trait_identifier.name.to_string();
+
+                        // Retrieve, from the trait definition, the expected method signature
+                        let contract_defining_trait = env.global_context.database.get_contract(&trait_identifier.contract_identifier)
+                            .map_err(|_e| CheckErrors::NoSuchContract(trait_identifier.contract_identifier.to_string()))?;
+                        let contract_context_defining_trait = contract_defining_trait.contract_context;
+
+                        // Retrieve the function that will be invoked
+                        let function_to_check = contract_context_to_check.lookup_function(function_name)
```

Comment:
Can this ever fail with `BadTraitImplementation`?  Don't we type-check a `contract-call?` before trying to evaluate it?

Reply:
Related to https://github.com/blockstack/blockstack-core/pull/1246#discussion_r377921009

Reply:
I'm trying to understand how this code can be reached.  If I define a trait in contract A, implement it in contract B, and try to call it within a `contract-call?` in contract C using contract B as the `impl`, then won't the type-checker be able to infer whether or not C's `contract-call?` is valid?

I can see code similar to this being necessary If we're issuing a `ContractCall` transaction, but that code that checks that both the contract and the `impl` exists would belong in the transaction evaluation logic, not here.

Reply:
Oh I think I see the blocker here. 
So we have 2 flavors of contracts that can be dynamically called:
1) Contracts that are compatible with a trait (as defined in the initial spec https://github.com/blockstack/blockstack-core/issues/1168)
2) Contracts that are explicitly declaring that they will be implementing a given trait, using the `impl-trait` statement. This flavor was added on the fly during the implementation of the spec, and is introduced in the description of this PR, but not is not mentioned in the initial spec.

In the snippet above we have these 2 branches being handled:
```
if contract_context_to_check.is_explicitely_implementing_trait
 // expedite flavor 2 - the analysis did all the work already
else
 // handle flavor 1
```

Reply:
I'm not sure this is a good idea.  It looks like it's imposing what should be a one-time analysis cost paid at publish time onto each runtime invocation of a `contract-call?`.

Reply:
I'm also not sure how safe this is.  Doesn't this violate point (2) on the "Preventing Cycles" section of #1168? 

```
Trait parameter <A> would not satisfy the <B> trait parameter in interface A — i.e., trait parameters are exact, even if a trait would always satisfy another trait.
```

Reply:
So, even in an 'accidental trait implementation', you can't use this get around cycle prevention -- ultimately, you'd need to be able to define a trait with a circular reference, which you still can't do. Rather, it just lets you perform the _outermost_ contract-call (i.e., the transaction itself) with a contract that implicitly implements the trait.

For this, though, I think these checks _should_ be moved to the outermost contract-call, precisely to avoid the cost of performing these checks on every dynamic contract-call (though, the cost of the check has the same asymptotic behavior as the contract-call?'s argument type checks in any case). However, I think that for the purposes of this PR, doing the check here is fine, but we should log an issue for moving the enforcement of this earlier.

Reply:
Thanks for clarifying!